app/controllers/spree/affirm_controller.rb

```diff
@@ -0,0 +1,37 @@
+module Spree
+  class AffirmController < Spree::StoreController
+    protect_from_forgery except: [:confirm]
+
+    def confirm
+      order = current_order || raise(ActiveRecord::RecordNotFound)
+      checkout_token = affirm_params[:checkout_token]
+
+      if !checkout_token
+        return redirect_to checkout_state_path(order.state), notice: "Invalid order confirmation data passed in"
+      end
+
+      if order.complete?
+        return redirect_to spree.order_path(order), notice: "Order is already in complete state"
+      end
+
+      affirm_checkout = Solidus::Affirm::Checkout.new(token: checkout_token)
+
+      if affirm_checkout.valid?
+        affirm_checkout.save
+        order.payments.create!({
+          payment_method_id: affirm_params[:payment_method_id],
+          amount: order.total,
+          source: affirm_checkout
+        })
+        order.next
```

Comment:
Then some site have bad luck, since this is how the order state machine currently works. That's why we call next instead of assuming a state, or keep looping over next until it's complete. So site's implementing different states can still assume this is moving it to the next state. Whatever that next state is. 

In the 80/20 rule of things, this falls in the 20 IMHO. If there is a nice and clean way to solve this, I'm more then happy to discus that and make it so of course.